ufl/algorithms/compute_form_data.py

```diff
@@ -344,7 +346,38 @@ def compute_form_data(form,
     self.reduced_coefficients = sorted(reduced_coefficients_set,
                                        key=lambda c: c.count())
     self.num_coefficients = len(self.reduced_coefficients)
-    self.original_coefficient_positions = [i for i, c in enumerate(self.original_form.coefficients())
+
+    old_coefficients = list(self.original_form.coefficients())
+    extops_positions = {}
+    for i, c in reversed(list(enumerate(self.original_form.coefficients()))):
+        if isinstance(c, ExternalOperator):
+            e = old_coefficients.pop(i)
+            if e.derivatives == (0,) * len(e.ufl_operands):
+                # Set the extop_id as the position in the original_form of e and its dependency
+                extops_positions[e._count] = i
+
+    # Store all the external operators and their derivative multiindex
+    # turning up in the form (Differentiation may have introduced new external operators)
+    derivatives_dict = {}
+    extop_dict = {}
+    for e in self.reduced_coefficients:
+        if isinstance(e, ExternalOperator):
+            eid = extops_positions[e._extop_master._count]
+            if eid not in derivatives_dict.keys():
+                derivatives_dict[eid] = ()
+                extop_dict[eid] = ()
+            derivatives_dict[eid] += (e.derivatives,)
+            extop_dict[eid] += (e,)
+    self.external_operators = derivatives_dict
```

Comment:
This is used outside (in a parallel context) so it needs to be and `OrderedDict` or some other kind of consistently ordered data structure.